### PR TITLE
Fixes accidental override of file conf settings with defaults

### PR DESCRIPTION
**Pull request by trozet**
When config is read, the config file is parsed first. After these
settings are loaded, the CLI settings are parsed and override any
settings read from the file. This is fine, except for the CLI settings
may be defaulted, which will just override any specified config file
settings with defaults.

This patch addresses the problem by checking during config override if
the the configuration option has a default specified, and if so, ignores
overriding the current config option.

Additionally, this patch adds real default values for settings.

Signed-off-by: Tim Rozet <trozet@redhat.com>

### CODE REVIEW

**Comment by trozet** [comment]
@danwinship

File: go-controller/pkg/config/config.go [hunk 1]
```diff
@@ -854,7 +878,7 @@ func buildGatewayConfig(ctx *cli.Context, cli, file *config) error {
 		}
 	}
 	// And CLI overrides over config file and default values
-	if err := overrideFields(&Gateway, &cli.Gateway); err != nil {
+	if err := overrideFields(&Gateway, &cli.Gateway, nil); err != nil {
```
Comment (trozet) [inline go-controller/pkg/config/config.go]: some config savedGateway is just empty values and no real defaults, so I just pass nil for those. Maybe I should pass var instead in case in the future there is a default value assigned for these...

**Comment by dcbw** [comment]
> This is fine, except for the CLI settings
> may be defaulted, which will just override any specified config file
> settings with defaults.

Looks like the only places that the CLI settings could get defaulted is for RawClusterSubnets and for boolean values? I think we should fix the defaulting for RawClusterSubnets (by removing the Value:) assignment in CommonFlags. And for booleans I don't think we need to care about supporting the config file setting the value to TRUE and then the CLI overriding that back to FALSE. So for bools perhaps we just logically OR the values together in `overrideFields()`? Is that enough to fix the problem?

**Comment by trozet** [comment]
> > This is fine, except for the CLI settings
> > may be defaulted, which will just override any specified config file
> > settings with defaults.
> 
> Looks like the only places that the CLI settings could get defaulted is for RawClusterSubnets and for boolean values? I think we should fix the defaulting for RawClusterSubnets by removing the `Value:` assignment in CommonFlags since that already has the same value in `Default.RawClusterSubnets`. And for booleans I don't think we need to care about supporting the config file setting the value to TRUE and then the CLI overriding that back to FALSE. So for bools perhaps we just logically OR the values together in `overrideFields()`? Is that enough to fix the problem?

I would argue that is just masking a bigger problem that settings are missing default values. For example,
	cli.IntFlag{
		Name:        "mtu",
		Usage:       "MTU value used for the overlay networks (default: 1400)",
		Destination: &cliConfig.Default.MTU,
	},

Default value says 1400, but it has no default. We just assume here that the config will be set to 1400 by default by the config file when it gets deployed? We should probably add real defaults here to match, and then the code I added will ensure proper overriding occurs.

**Comment by dcbw** [comment]
> Default value says 1400, but it has no default. We just assume here that the config will be set to 1400 by default by the config file when it gets deployed? We should probably add real defaults here to match, and then the code I added will ensure proper overriding occurs.

@trozet not sure what you mean here. The "default" value mentioned there actually comes from the top-level Default structure in config.go. In this case, the way the code *should* work (though perhaps you've found that reality doesn't match specification) is that:

1. Default.MTU = 1400 (around line 37 of config.go)
2. cliConfig.Default.MTU = 0 (because the --mtu flag was not passed)
3. file Default.MTU = 0 (because nothing set it in the file)
4. `buildDefaultConfig()` calls `overrideFields(&Default, &file.Default)` but because file.Default is the empty value (0) the 1400 is left in Default.MTU.
5. `buildDefaultConfig()` calls `overrideFields(&Default, &cli.Default)` but because cli.Default is the empty value (0) the 1400 is left in Default.MTU.
6. and all the code looks in that top-level Default.MTU value and finds 1400

Basically, the top-level structures around lines 20 - 100 are (a) filled with the default values and (b) what the code actually looks at when all config is handled. Then as other config comes in (file, CLI, env vars, special files from kubernetes) those override the top-level structure values in a defined order.

This is quite probably confusing and could be documented better. But the end goal (however it gets done) is to have a precedence of defaults < kube SA < env-vars < config file < CLI.

**Comment by trozet** [comment]
Yeah I mean the CLI settings themselves have no defaults. My latest patch sets them. I think this is a lot less ambiguous than the previous behavior. The problem is the configOverride method assumes what values are invalid. For example, it wont override if a setting is 0. What if a setting was added where 0 was a valid value? It also wont override if the boolean is false. What if a new setting was default true and we wanted to set it to false? That's why it is better to determine defaults in the settings themselves.

I'll add a follow up commit that will remove the checks for 0, true, etc from config override as they are not necessary.

File: go-controller/pkg/config/config.go [hunk 2]
```diff
@@ -757,54 +774,60 @@ func setOVSExternalID(exec kexec.Interface, key, value string) error {
...
 		return err
 	}
 
 	// And CLI overrides over config file and default values
-	if err := overrideFields(&Kubernetes, &cli.Kubernetes); err != nil {
+	if err := overrideFields(&Kubernetes, &cli.Kubernetes, &savedKubernetes); err != nil {
 		return err
 	}
 
+	// Grab default values from OVS external IDs
```
Comment (trozet) [inline go-controller/pkg/config/config.go]: Note I moved this down here, so that if the behavior is if the booleans are set to detect from OVS, they will override all other settings. Not sure if that is the desired behavior or not.

**Comment by trozet** [comment]
@dcbw take a look at the latest changes and see what you think. The checks with the switch statement for specific types are gone. Now it just compares against default values.

**Comment by trozet** [comment]
@danwinship noticed the latest patch isn't working, hold while I investigate it

**Comment by trozet** [comment]
Looks like the patch is working as expected. The issue was CLI arguments being passed via CNO setting cluster-subnets to nothing:
exec /usr/bin/ovnkube --init-node ovn-control-plane --cluster-subnets '' --k8s-service-cidr '' --k8s-apiserver https://192.168.1.5:11337

**Comment by danwinship** [comment]
Though it's reasonable to expect that in

    ovnkube --init-node --config-file=/blah/blah --cluster-subnets=10.0.0.0/8

the `--cluster-subnets` arg overrides the config file, while in

    ovnkube --init-node --cluster-subnets=10.0.0.0/8 ...  --config-file=/blah/blah

the config file overrides the `--cluster-subnets` arg

**Comment by trozet** [comment]
> Though it's reasonable to expect that in
> 
> ```
> ovnkube --init-node --config-file=/blah/blah --cluster-subnets=10.0.0.0/8
> ```
> 
> the `--cluster-subnets` arg overrides the config file, while in
> 
> ```
> ovnkube --init-node --cluster-subnets=10.0.0.0/8 ...  --config-file=/blah/blah
> ```
> 
> the config file overrides the `--cluster-subnets` arg

I don't think so. That then implies ordering of arguments matters, which I don't think is the case today. IMO we should just make it known that CLI takes precedence over config file (should be in the usage) and not add more caveats (there's already too many ways to determine config). If others feel strongly the ordering should matter, we can take that up in another PR.

**Comment by dcbw** [comment]
> Though it's reasonable to expect that in
> 
> ```
> ovnkube --init-node --config-file=/blah/blah --cluster-subnets=10.0.0.0/8
> ```
> 
> the `--cluster-subnets` arg overrides the config file, while in
> 
> ```
> ovnkube --init-node --cluster-subnets=10.0.0.0/8 ...  --config-file=/blah/blah
> ```
> 
> the config file overrides the `--cluster-subnets` arg

@danwinship current implementation has a clearly defined override, where CLI always overrides everything (config file, env args, kubernetes stuff, etc).

File: go-controller/pkg/config/config.go (hunk 2 above)
Comment (dcbw) [reply go-controller/pkg/config/config.go]: @girishmg will this be OK with you? We don't use the OVS external IDs, so I think this would only affect your setups.

**Comment by danwinship** [comment]
lgtm

**Comment by trozet** [comment]
@girishmg PTAL
